Sources/StreamChat/Repositories/MessageRepository.swift

```diff
@@ -0,0 +1,159 @@
+    case messageDoesNotExist
+    case messageIsNotPending
+    case messageDoesNotHaveValidChannel
+    case failedToSendMessage
+}
+
+class MessageRepository {
+    let database: DatabaseContainer
+    let apiClient: APIClient
+
+    init(database: DatabaseContainer, apiClient: APIClient) {
+        self.database = database
+        self.apiClient = apiClient
+    }
+
+    func sendMessage(
+        with messageId: MessageId,
+        completion: @escaping (Result<ChatMessage, MessageRepositoryError>) -> Void
+    ) {
+        // Check the message with the given id is still in the DB.
+        database.backgroundReadOnlyContext.perform { [weak self] in
+            guard let dto = self?.database.backgroundReadOnlyContext.message(id: messageId) else {
+                log.error("Trying to send a message with id \(messageId) but the message was deleted.")
+                completion(.failure(.messageDoesNotExist))
+                return
+            }
+
+            // Check the message still have `pendingSend` state.
+            guard dto.localMessageState == .pendingSend else {
+                log.info("Skipping sending message with id \(dto.id) because it doesn't have `pendingSend` local state.")
+                completion(.failure(.messageIsNotPending))
+                return
+            }
+
+            guard let channelDTO = dto.channel, let cid = try? ChannelId(cid: channelDTO.cid) else {
+                log.info("Skipping sending message with id \(dto.id) because it doesn't have a valid channel.")
+                completion(.failure(.messageDoesNotHaveValidChannel))
+                return
+            }
+
+            let requestBody = dto.asRequestBody() as MessageRequestBody
+
+            // Change the message state to `.sending` and the proceed with the actual sending
+            self?.database.write({
+                let messageDTO = $0.message(id: messageId)
+                messageDTO?.localMessageState = .sending
+
+            }, completion: { error in
+                if let error = error {
+                    log.error("Error changing localMessageState message with id \(messageId) to `sending`: \(error)")
+                    self?.markMessageAsFailedToSend(id: messageId) {
+                        completion(.failure(.failedToSendMessage))
+                    }
+                    return
+                }
+
+                let endpoint: Endpoint<MessagePayload.Boxed> = .sendMessage(cid: cid, messagePayload: requestBody)
+                self?.apiClient.request(endpoint: endpoint) {
+                    switch $0 {
+                    case let .success(payload):
+                        self?.saveSuccessfullySentMessage(cid: cid, message: payload.message) { message in
+                            if let message = message {
+                                completion(.success(message))
+                            } else {
+                                completion(.failure(.failedToSendMessage))
```

Comment:
Some logging would be nice here too.

Reply:
Logging is produced by `saveSuccessfullySentMessage`